beacon-chain/p2p/discovery.go

```diff
@@ -345,78 +339,116 @@ func (s *Service) listenForNewNodes() {
 				}
 			}()
 		}
 	}
 }
 
-// findAndDialPeers finds new peers until `targetPeerCount` is reached,
-// and dials them in batches of `maxConcurrentDials` at a time.
-// If the context is cancelled during the search, it stops looking for new peers,
-// but still dials new found peers. In this case, it returns the context error.
-func (s *Service) findAndDialPeers(ctx context.Context, targetPeerCount uint, maxConcurrentDials int) error {
-	for missingPeerCount := s.missingPeerCount(targetPeerCount); missingPeerCount > 0; {
-		// Create an discovery iterator to find new peers.
-		iterator := s.dv5Listener.RandomNodes()
-
-		// The peer crawling stops when enough peers are found in all defective subnets,
-		// or when `batchPeriod` is reached.
-		// It is useful to dial peers even if not all peers in defective subnets
-		// are found.
-		ctx, cancel := context.WithTimeout(ctx, batchPeriod)
-		defer cancel()
-
-		// `iterator.Next` can block indefinitely. `iterator.Close` unblocks it.
-		// So it is important to close the iterator when the context is done to ensure
-		// that the search does not hang indefinitely.
-		go func() {
-			<-ctx.Done()
-			iterator.Close()
-		}()
+// FindAndDialPeersWithSubnets ensures that our node is connected to enough peers.
+// If, the threshold is met, then this function immediately returns.
+// Otherwise, it searches for new peers and dials them.
+// If `ctx“ is canceled while searching for peers, search is stopped, but new found peers are still dialed.
+// In this case, the function returns an error.
+func (s *Service) findAndDialPeers(ctx context.Context) error {
+	// Restrict dials if limit is applied.
+	maxConcurrentDials := math.MaxInt
+	if flags.MaxDialIsActive() {
+		maxConcurrentDials = flags.Get().MaxConcurrentDials
+	}
 
-		// Crawl the network for peers subscribed to the defective subnets.
-		nodeByNodeID := make(map[enode.ID]*enode.Node)
-		for missingPeerCount > 0 && iterator.Next() {
-			if ctx.Err() != nil {
-				// If the context is done, we stop iterating.
-				break
-			}
-			// Iterate on peers and skip those that do not match the filter.
-			node := iterator.Node()
-			if !s.filterPeer(node) {
-				continue
+	missingPeerCount := s.missingPeerCount(s.cfg.MaxPeers)
+	for missingPeerCount > 0 {
+		// Stop the search/dialing loop if the context is canceled.
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+
+		peersToDial, err := func() ([]*enode.Node, error) {
+			ctx, cancel := context.WithTimeout(ctx, batchPeriod)
+			defer cancel()
+
+			peersToDial, err := s.findPeers(ctx, missingPeerCount)
+			if err != nil && !errors.Is(err, context.DeadlineExceeded) {
+				return nil, errors.Wrap(err, "find peers")
 			}
 
-			// Remove duplicates, keeping the node with higher seq.
-			existing, ok := nodeByNodeID[node.ID()]
-			if ok && existing.Seq() > node.Seq() {
-				continue
+			return peersToDial, nil
+		}()
+
+		if err != nil {
+			return errors.Wrap(err, "find peers")
```

Comment:
peersToDial returns an error here it will stutter, "find peers, find peers, some error"